File: clue/facts.cpp
#include "facts.h"
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std; 

facts::facts(vector<string> turnOrder)
{
	// every card should have the possibility of being with every person, so
	// need to initialize facts to represent this
	for (auto card : allCards)
	{
		vector<pair<string, string>> temp;
		for (auto person : turnOrder)
		{
			temp.push_back({ person, card });
		}
        temp.push_back({ "envelope", card });
		myFacts.push_back(temp);
	}
}

void facts::addFact(vector<pair<string, string>> possibilities)
{
	myFacts.push_back(possibilities);	
}

void facts::eliminate(vector<vector<string>>& cards, vector<string> names)
{
	// look through myFacts and check for redundancy against all others
    vector<vector<string>> confirmed = cards;
    bool changed = true;

    while (changed)
    {
        changed = false;

        // collect any new unit clauses as confirmed facts
        for (vector<pair<string, string>>& clause : myFacts)
        {
            if (clause.size() == 1)
            {
                vector<string> factVec{ clause[0].first, clause[0].second};
                if (find(confirmed.begin(), confirmed.end(), factVec) == confirmed.end())
                {
                    confirmed.push_back(factVec);
                    cards.push_back(factVec);
                    changed = true;
                }
            }
        }

        // need to check for players where you know all their cards and eliminate facts
        int numCards = 18 / names.size();
        vector<int> count(names.size(), 0);
        for (vector<string> fact : confirmed)
        {
            for (int i = 0; i < names.size(); ++i)
            {
                if (fact[0] == names[i])
                {
                    ++count[i];
                }
            }
        }

        for (int i = 0; i < count.size(); ++i)
        {
            // if we know all of their cards already...
            if (numCards == count[i])
            {
                // eliminate anything that isn't part of confirmed with their name, (P, X)
                for (vector<pair<string, string>>& clause : myFacts)
                {
                    int before = clause.size();
                    for (auto it = clause.begin(); it != clause.end(); )
                    {
                        bool isInConfirmed = false;
                        for (auto fact : confirmed)
                        {
                            if (it->first == fact[0] && it->second == fact[1])
                            {
                                isInConfirmed = true;
                            }
                        }

                        if (!isInConfirmed && it->first == names[i])
                        {
                            // erase returns the next iterator
                            it = clause.erase(it);
                        }
                        else
                        {
                            ++it;
                        }
                        
                    }

                    if (clause.size() != before)
                    {
                        changed = true;
                    }
                }
            }
        }

        // go through every confirmed fact (P,C)
        for (vector<string>& fact : confirmed)
        {
            string P = fact[0];
            string C = fact[1];
            pair<string, string> target{ P, C };

            // remove every clause that contains (P,C) since it's satisfied already
            for (auto it = myFacts.begin(); it != myFacts.end(); )
            {
                if (find(it->begin(), it->end(), target) != it->end())
                {
                    // dont want to remove the statements of length 1
                    if (it->size() > 1)
                    {
                        it = myFacts.erase(it);
                        changed = true;
                    }
                    else
                    {
                        ++it;
                    }
                }
                else
                {
                    ++it;
                }
            }

            // strip out any other (X,C) from all other clauses
            for (vector<pair<string, string>>& clause : myFacts)
            {
                // Only remove pairs for the same card BUT a different person
                int before = clause.size();
                clause.erase(remove_if(clause.begin(), clause.end(),
                    [&](const pair<string, string>& somePair) {
                        return somePair.second == C && somePair.first != P;
                    }), clause.end());
                if (clause.size() != before)
                {
                    changed = true;
                }
            }
        }
    }
	return;
}

File: web-clue/game/Facts.js
const { ALL_CARDS } = require('./globals');

class Facts {
    constructor(turnOrder = []) {
        this.myFacts = [];
        
        // Initialize facts - every card should have the possibility of being with every person
        for (const card of ALL_CARDS) {
            const possibilities = [];
            for (const person of turnOrder) {
                possibilities.push({ player: person, card: card });
            }
            possibilities.push({ player: "envelope", card: card });
            this.myFacts.push(possibilities);
        }
    }
    
    addFact(possibilities) {
        // Convert from array of [player, card] pairs to objects
        const factPossibilities = possibilities.map(pair => ({
            player: pair[0],
            card: pair[1]
        }));
        this.myFacts.push(factPossibilities);
    }
    
    eliminate(cards, playerNames) {
        const initialCopy = [...cards];
        const confirmed = [...cards];
        let changed = true;
        
        while (changed) {
            changed = false;
            
            // Collect any new unit clauses as confirmed facts
            for (const clause of this.myFacts) {
                if (clause.length === 1) {
                    const factArray = [clause[0].player, clause[0].card];
                    const exists = confirmed.some(card => 
                        card[0] === factArray[0] && card[1] === factArray[1]
                    );
                    
                    if (!exists) {
                        confirmed.push(factArray);
                        cards.push(factArray);
                        changed = true;
                    }
                }
            }
            
            // Check for players where we know all their cards
            const numCardsPerPlayer = Math.floor(18 / playerNames.length);
            const playerCardCounts = new Array(playerNames.length).fill(0);
            
            for (const fact of confirmed) {
                const playerIndex = playerNames.indexOf(fact[0]);
                if (playerIndex !== -1) {
                    playerCardCounts[playerIndex]++;
                }
            }
            
            for (let i = 0; i < playerCardCounts.length; i++) {
                if (numCardsPerPlayer === playerCardCounts[i]) {
                    // Eliminate anything that isn't part of confirmed with their name
                    for (const clause of this.myFacts) {
                        const sizeBefore = clause.length;
                        
                        for (let j = clause.length - 1; j >= 0; j--) {
                            const possibility = clause[j];
                            
                            if (possibility.player === playerNames[i]) {
                                const isConfirmed = confirmed.some(card => 
                                    card[0] === possibility.player && card[1] === possibility.card
                                );
                                
                                if (!isConfirmed) {
                                    clause.splice(j, 1);
                                }
                            }
                        }
                        
                        if (clause.length !== sizeBefore) {
                            changed = true;
                        }
                    }
                }
            }
            
            // Process each confirmed fact
            for (const fact of confirmed) {
                const playerName = fact[0];
                const cardName = fact[1];
                const target = { player: playerName, card: cardName };
                
                // Remove clauses that contain this fact (they're satisfied)
                for (let i = this.myFacts.length - 1; i >= 0; i--) {
                    const clause = this.myFacts[i];
                    
                    if (clause.some(p => p.player === target.player && p.card === target.card)) {
                        if (clause.length > 1) {
                            this.myFacts.splice(i, 1);
                            changed = true;
                        }
                    }
                }
                
                // Remove other players having this same card from all clauses
                for (const clause of this.myFacts) {
                    const sizeBefore = clause.length;
                    
                    for (let j = clause.length - 1; j >= 0; j--) {
                        const possibility = clause[j];
                        
                        if (possibility.card === cardName && possibility.player !== playerName) {
                            clause.splice(j, 1);
                        }
                    }
                    
                    if (clause.length !== sizeBefore) {
                        changed = true;
                    }
                }
            }
        }
        
        // Return newly discovered cards
        const newDeductions = [];
        for (const card of cards) {
            const inInitial = initialCopy.some(initialCard => 
                initialCard[0] === card[0] && initialCard[1] === card[1]
            );
            
            if (!inInitial) {
                newDeductions.push({
                    playerName: card[0],
                    cardName: card[1],
                    message: `Discovered ${card[0]} has ${card[1]}`
                });
            }
        }
        
        return newDeductions;
    }
}

module.exports = Facts;